# <font color = 'green'> Computer Science Final Project </font>

## Andreose Eugenio

### n. matricola : 834332

#### Data Science a.a. 2017/2018

<font color = 'green'>**Starting from *The Movies Database* (TMDB) sample datasets**</font>

In [342]:
import pandas as pd
import numpy as np
import ast
from datetime import datetime
import tldextract

In [343]:
movies = pd.read_csv("C:/Users/User/Desktop/DATA SCIENCE/COMPUTER SCIENCE/FINAL_PROJECT/tmdb_5000_movies.csv")
credits = pd.read_csv("C:/Users/User/Desktop/DATA SCIENCE/COMPUTER SCIENCE/FINAL_PROJECT/tmdb_5000_credits.csv")

## <font color = 'green'>**(1) For each movie, compute the number of cast members**</font>

The information about cast members are located in "*tmdb_5000_credits*" dataset. The difficulty here is that all the characters in a cast, related to a specific movie, are contained in a dictionary hidden by a string structure.
In order to count the number of characters that appear into each , we must transform each Series of string in a dictionary, and count the number of character for each movie.

In [344]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [345]:
credits.cast.head()

0    [{"cast_id": 242, "character": "Jake Sully", "...
1    [{"cast_id": 4, "character": "Captain Jack Spa...
2    [{"cast_id": 1, "character": "James Bond", "cr...
3    [{"cast_id": 2, "character": "Bruce Wayne / Ba...
4    [{"cast_id": 5, "character": "John Carter", "c...
Name: cast, dtype: object

In [346]:
type(credits.cast[0])

str

In [347]:
len(credits.cast)

4803

In [349]:
a = np.zeros(4803, dtype= dict)
credits['cast_number'] = pd.Series(np.zeros(len(credits)), index=credits.index)

In [350]:
for i in range(0,len(credits)):
    credits.iloc[i,4] =len(ast.literal_eval(credits.cast[i]))
credits["cast_number"] = credits["cast_number"].astype(int)

In [354]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  cast_number  
0  [{"credit_id": "52fe48009251416c750aca23", "de...           83  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...           34  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...           83  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...          158  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...           27

## <font color = 'green'>**(2) How many movies do not have an homepage?** </font>

The information about movies homepage is located in "*tmdb_5000_movies*" dataset.

In [355]:
movies["homepage"].isnull().sum()

3091

As we could see from the above output, $3091$ movies don't possess an own homepage.

## <font color = 'green'>** (3) For each year, how many movies do not have a homepage?**</font>

In [356]:
type(movies.release_date[0])

str

As we could see from the output above, the comulm "release_date" is a string. We need to convert it in a pandas date format.

In [357]:
movies['release_date'] = pd.to_datetime(movies['release_date'])

Now we have to consider only the year of the relase date. Pandas can do that with attribute "dt" and "year". We'll add a column named "year" that indicate only the year of release. In a second moment I'll encode with a integer type the counting and Na (movies with no release date avaible) with 0.

In [358]:
movies['year'] = movies['release_date'].dt.year
movies['year'] = movies['year'].fillna(0).astype(int)

In [359]:
movies[movies.homepage.isnull()].groupby('year').agg({"homepage":sum,"year":"count"}).head()

homepage  year
year                
0          NaN     1
1916       NaN     1
1925       NaN     1
1927       NaN     1
1929       NaN     2

## <font color = 'green'>(4) Extract the domain of each homepage. </font>

With "tldextract" package it's possible to extract many element of an URL. In particular now we'll consider the domain of an homepage.

In [360]:
movies.homepage[0]

'http://www.avatarmovie.com/'

In [361]:
tldextract.extract(movies.homepage[0]).domain

'avatarmovie'

As similar as the above procedure, we'll now add a column with a reference of the domain of each movies's hompage (if the hompage exist).

In [362]:
def domain (c):
    try :
        domain = tldextract.extract(c).domain
    except:
        domain = ""
    return domain

movies['Domain']=movies['homepage'].apply(domain)

In [363]:
movies[['original_title','homepage', 'Domain']].head()

original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                       homepage              Domain  
0                   http://www.avatarmovie.com/         avatarmovie  
1  http://disney.go.com/disneypictures/pirates/                  go  
2   http://www.sonypictures.com/movies/spectre/        sonypictures  
3            http://www.thedarkknightrises.com/  thedarkknightrises  
4          http://movies.disney.com/john-carter              disney

## <font color = 'green'> (5) Extract a set of normalized tables. That is, each entry of a normalized table must contain exactly one value (not a list or a dictionary).</font> 

The columns to normalize are :

* Movies : 
        + Genres
        + Keywords
        + Production_companies
        + Production_countries
        + Spoken_languages
        
* Credits :
        + Cast
        + Crew

For this task I have to create a new DataFrame in which each raw refers to a single value. In adiction I have to sill taking into account the movie id. 
I'have decided to create a function that with a DataFrame, a specific column and Id, returns the normalized table.

In [53]:
def to_table (DataFrame, column, id_col) :
    iniz = pd.DataFrame()
    for i in range(0,len(DataFrame)):
        aDF = pd.read_json(DataFrame[column][i])
        aDF[id_col] = DataFrame[id_col][i]
        iniz = pd.concat([iniz,aDF],0)
    return iniz

I'll now apply this function to each column of both Dataframes that has to be normalized:

#### Genres

In [54]:
m_genres = to_table(movies, 'genres','id')
m_genres.head()

id             name
0.0  19995           Action
1.0  19995        Adventure
2.0  19995          Fantasy
3.0  19995  Science Fiction
0.0    285        Adventure

#### Keywords

In [55]:
m_keyword = to_table(movies, 'keywords','id')
m_keyword.head()

id           name
0.0  19995  culture clash
1.0  19995         future
2.0  19995      space war
3.0  19995   space colony
4.0  19995        society

#### Production_companies

In [56]:
m_prodcomp = to_table(movies, 'production_companies','id')
m_prodcomp.head()

id                                    name
0.0  19995                 Ingenious Film Partners
1.0  19995  Twentieth Century Fox Film Corporation
2.0  19995                      Dune Entertainment
3.0  19995                Lightstorm Entertainment
0.0    285                    Walt Disney Pictures

#### Production_countries

In [57]:
m_prodcont = to_table(movies, 'production_countries','id')
m_prodcont.head()

id iso_3166_1                      name
0.0   19995         US  United States of America
1.0   19995         GB            United Kingdom
0.0     285         US  United States of America
0.0  206647         GB            United Kingdom
1.0  206647         US  United States of America

#### Spoken_lenaguages

In [58]:
m_spoklang = to_table(movies, 'spoken_languages','id')
m_spoklang.head()

id iso_639_1      name
0.0   19995        en   English
1.0   19995        es   Español
0.0     285        en   English
0.0  206647        fr  Français
1.0  206647        en   English

#### Cast

In [59]:
c_cast = to_table(credits, 'cast','movie_id')
c_cast.head()

cast_id            character                 credit_id  gender       id  \
0.0    242.0           Jake Sully  5602a8a7c3a3685532001c9a     2.0  65731.0   
1.0      3.0              Neytiri  52fe48009251416c750ac9cb     1.0   8691.0   
2.0     25.0  Dr. Grace Augustine  52fe48009251416c750aca39     1.0  10205.0   
3.0      4.0        Col. Quaritch  52fe48009251416c750ac9cf     2.0  32747.0   
4.0      5.0         Trudy Chacon  52fe48009251416c750ac9d3     1.0  17647.0   

     movie_id                name  order  
0.0     19995     Sam Worthington    0.0  
1.0     19995         Zoe Saldana    1.0  
2.0     19995    Sigourney Weaver    2.0  
3.0     19995        Stephen Lang    3.0  
4.0     19995  Michelle Rodriguez    4.0

#### Crew

In [60]:
c_crew = to_table(credits, 'crew','movie_id')
c_crew.head()

credit_id  department  gender      id  \
0.0  52fe48009251416c750aca23     Editing     0.0  1721.0   
1.0  539c47ecc3a36810e3001f87         Art     2.0   496.0   
2.0  54491c89c3a3680fb4001cf7       Sound     0.0   900.0   
3.0  54491cb70e0a267480001bd0       Sound     0.0   900.0   
4.0  539c4a4cc3a36810c9002101  Production     1.0  1262.0   

                          job  movie_id               name  
0.0                    Editor     19995  Stephen E. Rivkin  
1.0         Production Design     19995        Rick Carter  
2.0            Sound Designer     19995  Christopher Boyes  
3.0  Supervising Sound Editor     19995  Christopher Boyes  
4.0                   Casting     19995          Mali Finn

## <font color = 'green'>(6) For each movie, compute the gross margin (difference between revenue and budget) </font>

In [364]:
movies["gross_margin"] =  movies.revenue - movies.budget
movies[['original_title','revenue', 'budget','gross_margin']].head()

original_title     revenue     budget  \
0                                    Avatar  2787965087  237000000   
1  Pirates of the Caribbean: At World's End   961000000  300000000   
2                                   Spectre   880674609  245000000   
3                     The Dark Knight Rises  1084939099  250000000   
4                               John Carter   284139100  260000000   

   gross_margin  
0    2550965087  
1     661000000  
2     635674609  
3     834939099  
4      24139100

## <font color = 'green'>(7) For each movie, compute the number of crew members </font>

Due to obtain the number of crew members for each movie, I have to *join* the following two tables:
* **Crew** that is the column normalized before that contains crew members.
* **Credits** that contains movies title.

The *key* used to join those two table is the *movie_id*. 


In [365]:
c_crew_gb = c_crew.groupby("movie_id")[["credit_id"]].count()
c_crew_gb["movie_id"] = c_crew_gb.index
c_crew_gb.head()

credit_id  movie_id
movie_id                     
5                88         5
11               20        11
12              104        12
13               93        13
14              109        14

In [366]:
credits[['title', 'movie_id']].head()

title  movie_id
0                                    Avatar     19995
1  Pirates of the Caribbean: At World's End       285
2                                   Spectre    206647
3                     The Dark Knight Rises     49026
4                               John Carter     49529

In [367]:
pd.merge(credits[['title','movie_id']],
        c_crew_gb,
        on = 'movie_id',
        how = 'right').head()

title  movie_id  credit_id
0                                    Avatar     19995        153
1  Pirates of the Caribbean: At World's End       285         32
2                                   Spectre    206647        155
3                     The Dark Knight Rises     49026        217
4                               John Carter     49529        132

## <font color = 'green'>(8) For each movie, compute the number of directors </font>

We'll proceede in the same way as before:

In [368]:
c_crew_dir = c_crew[c_crew['job'] == 'Director'].groupby('movie_id')[["credit_id"]].count()
c_crew_dir['movie_id'] = c_crew_dir.index
c_crew_dir.head()

credit_id  movie_id
movie_id                     
5                 4         5
11                1        11
12                1        12
13                1        13
14                1        14

In [369]:
pd.merge(credits[['title','movie_id']],
        c_crew_dir,
        on = 'movie_id',
        how = 'right').head()

title  movie_id  credit_id
0                                    Avatar     19995          1
1  Pirates of the Caribbean: At World's End       285          1
2                                   Spectre    206647          1
3                     The Dark Knight Rises     49026          1
4                               John Carter     49529          1

## <font color = 'green'>(9) For each language, compute the number of movies where such language is spoken </font>

Using the normalized table refearing to the spoken languge of movies:

In [370]:
m_spoklang.groupby("name")[['id']].count().head()

id
name                
                  54
??????             1
Afrikaans          7
Bahasa indonesia   2
Bamanankan         1

## <font color = 'green'>(10) For each company and each decade, compute the overall revenue</font>

#### Overall Revenue for each *Company*:

In [371]:
movies[["revenue","id"]].head()

revenue      id
0  2787965087   19995
1   961000000     285
2   880674609  206647
3  1084939099   49026
4   284139100   49529

In [372]:
m_prodcomp.head()

id                                    name
0.0  19995                 Ingenious Film Partners
1.0  19995  Twentieth Century Fox Film Corporation
2.0  19995                      Dune Entertainment
3.0  19995                Lightstorm Entertainment
0.0    285                    Walt Disney Pictures

In [373]:
pd.merge(m_prodcomp,
        movies[['revenue','id']],
        on = 'id',
        how = 'right').groupby('name').agg({'revenue' : 'sum'}).head()

revenue
name                                     
"DIA" Productions GmbH & Co. KG  44350926
1.85 Films                          98017
10 West Studios                         0
100 Bares                        57965843
1019 Entertainment                2801508

#### Overall Revenue for each *Decade* :

First of all, i'll create a new coloumn refering to wich decade each movie belongs to. Movies without information about *relase date* (encoded with 0 value) wil be excluded.

In [374]:
movies['year_decade'] = pd.cut(movies['year'], np.arange(movies.year[movies.year != 0 ].min(), movies['year'].max() +1 , 10))

In [375]:
movies[['id','year','revenue','year_decade']].head()

id  year     revenue   year_decade
0   19995  2009  2787965087  (2006, 2016]
1     285  2007   961000000  (2006, 2016]
2  206647  2015   880674609  (2006, 2016]
3   49026  2012  1084939099  (2006, 2016]
4   49529  2012   284139100  (2006, 2016]

In a second moment, the overall revenue for eache decade has been computed.

In [376]:
movies.groupby('year_decade').agg({'revenue' : 'sum'}).head()

revenue
year_decade             
(1916, 1926]    22000000
(1926, 1936]    36427447
(1936, 1946]  1199672787
(1946, 1956]   315039841
(1956, 1966]  1700766586

## <font color = 'green'>(11) For each decade, compute the company with maximum revenue</font>

In [381]:
m_prodcomp.head()

id                                    name
0.0  19995                 Ingenious Film Partners
1.0  19995  Twentieth Century Fox Film Corporation
2.0  19995                      Dune Entertainment
3.0  19995                Lightstorm Entertainment
0.0    285                    Walt Disney Pictures

In [384]:
movies[['year_decade','id','revenue']].head()

year_decade      id     revenue
0  (2006, 2016]   19995  2787965087
1  (2006, 2016]     285   961000000
2  (2006, 2016]  206647   880674609
3  (2006, 2016]   49026  1084939099
4  (2006, 2016]   49529   284139100

In [440]:
pd.merge(m_prodcomp,
              movies[['year_decade','id','revenue']],
              on = 'id',
              how = 'right').groupby('year_decade')['id','revenue','name'].agg({'revenue' : 'max'})

revenue            
                  id     revenue
year_decade                     
(1916, 1926]    3060    22000000
(1926, 1936]   65203     8500000
(1936, 1946]   60400   400176459
(1946, 1956]   56930    36000000
(1956, 1966]   91721   286214286
(1966, 1976]   72086   470654000
(1976, 1986]  218500   792910554
(1986, 1996]  426067   920100000
(1996, 2006]  333355  1845034188
(2006, 2016]  459488  2787965087

And that's the *Id* of the companies that have earned more  during each decade.

## <font color = 'green'>(12) In each year, how many movies have revenue smaller than the budget?</font>

In [377]:
movies[['title','year','revenue','budget']].head()

title  year     revenue     budget
0                                    Avatar  2009  2787965087  237000000
1  Pirates of the Caribbean: At World's End  2007   961000000  300000000
2                                   Spectre  2015   880674609  245000000
3                     The Dark Knight Rises  2012  1084939099  250000000
4                               John Carter  2012   284139100  260000000

In [379]:
movies[movies['revenue']<movies['budget']].groupby('year').size()

year
1927     1
1947     1
1948     1
1954     1
1962     1
1963     1
1964     2
1965     3
1966     1
1968     2
1969     1
1970     4
1971     1
1972     1
1974     1
1977     4
1978     1
1979     2
1980     5
1981    11
1982     1
1983     5
1984     9
1985     9
1986     9
1987     3
1988     4
1989     6
1990     4
1991    10
1992     9
1993    15
1994    15
1995    19
1996    32
1997    33
1998    39
1999    55
2000    53
2001    56
2002    71
2003    47
2004    61
2005    55
2006    63
2007    55
2008    64
2009    78
2010    65
2011    70
2012    52
2013    62
2014    59
2015    67
2016    26
dtype: int64